core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java

```diff
@@ -302,6 +303,9 @@ private void demote() throws Exception {
     // sockets in stopServing so that clients don't see NPEs.
     mJournalSystem.losePrimacy();
     stopMasterComponents();
+
+    ProcessUtils.dumpInformationOnFailover();
```

Comment:
This one might be controversial because i'm dumping after the master components have been killed. If there are errors in threads, the jstacks will not be able to catch that problem. 

We can choose to perform the dump before demoting, at the risk of demoting being a bit slower. WDYT? @yuzhu

Reply:
This is a bit of a dilemma. If I output here, I won't be able to catch it when there's for example a thread leak in the journal, because `JournalSystem.losePrimacy()` will probably kill those threads. Same argument for not being able to see the thread leaks in the master components or deadlocks etc.

However, if I dump these information BEFORE `losePrimacy()`,  the correctness is endangered (see comments above that `losePrimacy()` line) because we have to give up primacy ASAP. @tcrain WDYT?

Reply:
The part that has to be done fast is shutting down the RPC service so we don't serve new client requests. Maybe you can shut down that service first, then do the collect info before shutting down the other services.

Reply:
Per our discussion, I've changed dumping metrics and jstack to async and before closing the RPC/journal, so that we have hope for capturing more critical information.

Metrics dump must happen before `MetricsMaster.close()`, but for simplicity I did not add more mechanism to handle that (like a Latch). In other words, I chose simplicity over the guarantee, because:
1. The close sequence is close RPC/WebServer -> close Journal -> close MetricsMaster, so most likely the metrics dump happens before MetricsMaster is dead. Getting the metrics should be fast (whereas outputing into the file is the slow part). So long as this async thread gets CPU at all, I think we are fine.
2. In a more serious env, the metrics should be captured by Prometheus so that particular one is of low importance.